app/models/voucher.rb

```diff
@@ -1,15 +1,136 @@
 # frozen_string_literal: false
 
 class Voucher < ApplicationRecord
+  acts_as_paranoid
+
+  include CalculatedAdjustments
+
   belongs_to :enterprise
 
+  has_many :adjustments,
+           as: :originator,
+           class_name: 'Spree::Adjustment',
+           inverse_of: :voucher,
+           dependent: :nullify
+
   validates :code, presence: true, uniqueness: { scope: :enterprise_id }
 
+  before_validation :add_calculator
+
+  def self.adjust!(order)
+    return if order.nil?
+
+    # Find open Voucher Adjustment
+    return if order.vouchers.empty?
+
+    # We only support one voucher per order right now, we could just loop on vouchers
+    adjustment = order.vouchers.first
+
+    # Recalculate value
+    amount = adjustment.originator.compute_amount(order)
+
+    if order.additional_tax_total.positive?
+      handle_tax_excluded_from_price(order, amount)
+    else
+      handle_tax_included_in_price(order, amount)
+    end
+
+    # Move to closed state
+    adjustment.close
+  end
+
+  def self.handle_tax_excluded_from_price(order, amount)
+    voucher_rate = amount / order.total
+
+    # TODO: might need to use VoucherTax has originator (sub class of Voucher)
+    # Adding the voucher tax part
+    tax_amount = voucher_rate * order.additional_tax_total
+
+    adjustment = order.vouchers.first
+    adjustment_attributes = {
+      amount: tax_amount,
+      originator: adjustment.originator,
+      order: order,
+      label: "Tax #{adjustment.label}",
+      mandatory: false,
+      state: 'closed',
+      tax_category: nil,
+      included_tax: 0
+    }
+    order.adjustments.create(adjustment_attributes)
+
+    # Update the adjustment amount
+    amount = voucher_rate * (order.total - order.additional_tax_total)
+
+    adjustment.update_columns(
+      amount: amount,
+      updated_at: Time.zone.now
+    )
+  end
```

Comment:
Ideally all the calculation would be encapsulated in a `Calculator` but I don't think creating the tax adjustment belongs in one of them. So I kept it in the `Voucher` class.
That means we need to call `Voucher.adjust!` and we can't just rely on "Spree::Ajustment.update_adjustment!" ( It is call by the order `updater` )
https://github.com/openfoodfoundation/openfoodnetwork/blob/c34ced2f65a7ffed514200864d85b4b5ccef6da5/app/models/spree/adjustment.rb#L100-L117